### PR TITLE
1331: Optimize build times - use fmt as a static library

### LINKED ISSUE #1331: Figure out why build/test times increased
**What Needs to be Done?**

Research what happened here to solve the problem.

### PR DESCRIPTION
**Pull request by cz4rs**
experimental PR for #1331 
fixes #1331

### CODE REVIEW

**Comment by PhilMiller** [comment]
Could we make the include of the fmt headers via `-isystem` instead of `-I`, so that warnings inside it are suppressed?
Relatedly, build its directory with `-w` for the same reasons

**Comment by cz4rs** [comment]
> Could we make the include of the fmt headers via `-isystem` instead of `-I`, so that warnings inside it are suppressed?
> Relatedly, build its directory with `-w` for the same reasons

I have fixed the warning for now, but `-isystem` and `-w` should be superior - I will have a look at that.

**Comment by cz4rs** [comment]
A simple test with a cleared cache:
`time /usr/bin/cmake --build . --config Release --target install -- -j 10`

**develop**
```
real    9m21,086s
user    70m12,107s
sys     1m22,634s
```
**1331-optimize-fmt-use**
```
real    7m34,897s
user    55m55,678s
sys     1m15,743s
```

**Comment by cz4rs** [comment]
> > Could we make the include of the fmt headers via `-isystem` instead of `-I`, so that warnings inside it are suppressed?
> > Relatedly, build its directory with `-w` for the same reasons
> 
> I have fixed the warning for now, but `-isystem` and `-w` should be superior - I will have a look at that.
> 
> _sidenote: we use `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag}" PARENT_SCOPE)` to set the flags instead of "modern CMake" recommended `target_compile_options` which could also do the trick_

In the end, `os.h` header that contained problematic code was removed completely. I can still go for `-isystem` and `-w` if you think that it provides value.

File: lib/fmt/CMakeLists.txt [hunk 1]
```diff
@@ -1,17 +1,94 @@
...
+  set(headers ${${VAR}})
+  foreach (header ${ARGN})
+    set(headers ${headers} include/fmt/${header})
+  endforeach()
+  set(${VAR} ${headers} PARENT_SCOPE)
+endfunction()
+
+# Define the fmt library, its includes and the needed defines.
+add_headers(FMT_HEADERS chrono.h color.h compile.h core.h format.h format-inl.h
+                        locale.h ostream.h printf.h ranges.h)
```
Comment (JacobDomagala) [inline lib/fmt/CMakeLists.txt:74]: Do we want to add all the files? I think currently we only use core and format(h/inl.h/cc) and the rest seems optional

Comment (cz4rs) [reply lib/fmt/CMakeLists.txt:74]: Good idea, I'll try to cut out as much as I can.

File: lib/fmt/CMakeLists.txt [hunk 2]
```diff
@@ -1,17 +1,38 @@
...
-#  DESTINATION  include/fmt
-#)
+project(FMT CXX)
+
+if (NOT CMAKE_RUNTIME_OUTPUT_DIRECTORY)
+  set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+endif ()
+
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH}
+  "${CMAKE_CURRENT_SOURCE_DIR}/support/cmake")
```
Comment (JacobDomagala) [inline lib/fmt/CMakeLists.txt:8]: Do we need these (together with `lib/fmt/support/cmake/fmt-config.cmake.in`)?

Comment (JacobDomagala) [reply lib/fmt/CMakeLists.txt:8]: We're building the static lib so `CMAKE_RUNTIME_OUTPUT_DIRECTORY` shouldn't be used, and `fmt-config.cmake.in` is for (I think) generating packages which we also don't do (with fmt) so I think these should be removed

Comment (lifflander) [reply lib/fmt/CMakeLists.txt:8]: I think you are correct @JacobDomagala, unless we are exporting the configuration. Do we need to be doing that now that it's not just header only for apps that use VT.

**Review by lifflander — APPROVED** [review]

File: lib/fmt/CMakeLists.txt (hunk 2 above)
Comment (lifflander) [reply lib/fmt/CMakeLists.txt:8]: Once we work this out, I think this is ready to merge.

**Comment by JacobDomagala** [comment]
@cz4rs  I think we might want to disable `496 - vt:*TestRDMAHandleCollection/*test_rdma_handle_collection_1_proc_4 (Timeout)`  for now, it's been randomly timing out on OpenMPI lately

**Comment by cz4rs** [comment]
> @cz4rs I think we might want to disable `496 - vt:*TestRDMAHandleCollection/*test_rdma_handle_collection_1_proc_4 (Timeout)` for now, it's been randomly timing out on OpenMPI lately

Disabled, looks like it should be fine now.

File: lib/fmt/CMakeLists.txt (hunk 2 above)
Comment (cz4rs) [reply lib/fmt/CMakeLists.txt]: I have removed it all, no issues with sample project observed :+1:

**Review by JacobDomagala — APPROVED** [review]

**Pull request merged**